### PR TITLE
Switch graphula from an initial algebra to a final tag-less representation.

### PR DESCRIPTION
**Pull request by eborden**
This commit introduces MonadGraphulaFrontend and MonadGraphulaBackend, the two demarcations of functionality in Graphula. We can likely split this even more, but it isn't necessary now.

Most graphula graphs are very small and this switch would likely not be necessary. However recently we've been experimenting with graphula for generating large sets of arbitrary test data for benchmarking etc. The free representation is extremely problematic in this situation.

With profiling and benchmarking we can see this

### Free
```
	total time  =       12.38 secs   (12382 ticks @ 1000 us, 1 processor)
	total alloc = 23,944,694,912 bytes  (excludes profiling overheads)

COST CENTRE        MODULE                            SRC                                                    %time %alloc

>>=.\              Control.Monad.Trans.Free          src/Control/Monad/Trans/Free.hs:(229,39)-(231,44)       21.1   27.2
>>=                Control.Monad.Trans.Free          src/Control/Monad/Trans/Free.hs:(229,3)-(231,44)        18.2   16.6
fmap               Graphula                          src/Graphula.hs:331:1-48                                11.9   20.1
getOverhead        Criterion.Monad                   Criterion/Monad.hs:(47,1)-(56,12)                        6.9    0.0
getGCStats         Criterion.Measurement             Criterion/Measurement.hs:(46,1)-(48,16)                  4.9    0.0
>>=                Data.Vector.Fusion.Util           Data/Vector/Fusion/Util.hs:33:3-18                       3.7    3.7
fmap               Graphula                          src/Graphula.hs:317:1-58                                 2.8    6.5
transFreeT         Control.Monad.Trans.Free          src/Control/Monad/Trans/Free.hs:337:1-79                 2.6    3.4
fmap               Control.Monad.Trans.Free          src/Control/Monad/Trans/Free.hs:(209,3)-(211,42)         2.4    2.6
fmap               Control.Monad.Trans.Free          src/Control/Monad/Trans/Free.hs:(132,3)-(133,37)         1.9    1.8
iterT              Control.Monad.Trans.Free          src/Control/Monad/Trans/Free.hs:(309,1)-(313,21)         1.6    1.4
basicUnsafeWrite   Data.Vector.Primitive.Mutable     Data/Vector/Primitive/Mutable.hs:114:3-69                1.5    0.7
fmap.f'            Control.Monad.Trans.Free          src/Control/Monad/Trans/Free.hs:(210,5)-(211,42)         1.2    1.7
basicUnsafeIndexM  Data.Vector.Primitive             Data/Vector/Primitive.hs:215:3-75                        1.0    1.0
backendArbitrary.\ Graphula                          src/Graphula.hs:(232,24)-(234,10)                        1.0    0.5
innerProduct.\     Statistics.Matrix.Algorithms      Statistics/Matrix/Algorithms.hs:42:3-51                  1.0    0.8
fmap               Data.Vector.Fusion.Stream.Monadic Data/Vector/Fusion/Stream/Monadic.hs:(125,3)-(127,20)    0.9    1.1
```

```
time                 7.636 μs   (7.332 μs .. 7.960 μs)
                     0.989 R²   (0.983 R² .. 0.996 R²)
mean                 7.578 μs   (7.383 μs .. 7.929 μs)
std dev              863.0 ns   (559.6 ns .. 1.460 μs)
variance introduced by outliers: 89% (severely inflated)
             
benchmarking initial algebra/100
time                 3.864 ms   (3.709 ms .. 4.046 ms)
                     0.991 R²   (0.987 R² .. 0.996 R²)
mean                 3.932 ms   (3.879 ms .. 3.999 ms)
std dev              196.6 μs   (158.3 μs .. 260.0 μs)
variance introduced by outliers: 30% (moderately inflated)
             
benchmarking initial algebra/1000
time                 371.4 ms   (346.2 ms .. 401.8 ms)
                     0.999 R²   (0.997 R² .. 1.000 R²)
mean                 372.5 ms   (368.3 ms .. 376.2 ms)
std dev              5.980 ms   (0.0 s .. 6.390 ms)
variance introduced by outliers: 19% (moderately inflated)
```

### Tag-less
```
	total time  =       14.48 secs   (14483 ticks @ 1000 us, 1 processor)
	total alloc = 21,409,688,160 bytes  (excludes profiling overheads)

COST CENTRE            MODULE                            SRC                                                    %time %alloc

nodeEditWith           Graphula                          src/Graphula.hs:(422,1)-(426,10)                        16.6   15.9
replicateNodeFinal     Main                              bench/Main.hs:68:1-74                                   13.4   17.9
attemptsToInsertWith   Graphula                          src/Graphula.hs:(501,1)-(508,60)                        12.8   12.4
>>=                    Data.Vector.Fusion.Util           Data/Vector/Fusion/Util.hs:33:3-18                       6.3    8.2
insert                 Main                              bench/Main.hs:58:3-33                                    6.1    9.7
getOverhead            Criterion.Monad                   Criterion/Monad.hs:(47,1)-(56,12)                        5.9    0.0
getGCStats             Criterion.Measurement             Criterion/Measurement.hs:(46,1)-(48,16)                  4.1    0.0
newTFGen               System.Random.TF.Init             src/System/Random/TF/Init.hs:82:1-43                     2.7    8.8
attemptsToInsertWith.\ Graphula                          src/Graphula.hs:507:17-22                                2.4    3.5
basicUnsafeWrite       Data.Vector.Primitive.Mutable     Data/Vector/Primitive/Mutable.hs:114:3-69                2.1    1.5
generate               Test.QuickCheck.Gen               Test/QuickCheck/Gen.hs:(101,1)-(103,20)                  2.0    2.7
newQCGen               Test.QuickCheck.Random            Test/QuickCheck/Random.hs:80:1-31                        1.9    0.0
fmap                   Data.Vector.Fusion.Stream.Monadic Data/Vector/Fusion/Stream/Monadic.hs:(125,3)-(127,20)    1.8    2.3
logNode                Main                              bench/Main.hs:65:3-21                                    1.7    0.0
basicUnsafeIndexM      Data.Vector.Primitive             Data/Vector/Primitive.hs:215:3-75                        1.7    2.1
rSquare.p              Statistics.Regression             Statistics/Regression.hs:109:5-66                        1.6    1.6
innerProduct.\         Statistics.Matrix.Algorithms      Statistics/Matrix/Algorithms.hs:42:3-51                  1.5    1.7
kbnAdd                 Numeric.Sum                       Numeric/Sum.hs:(138,1)-(141,39)                          1.5    0.9
basicUnsafeIndexM      Data.Vector.Unboxed.Base          Data/Vector/Unboxed/Base.hs:283:841-899                  1.2    0.0
qr.\.\                 Statistics.Matrix.Algorithms      Statistics/Matrix/Algorithms.hs:29:34-51                 1.1    1.3
transpose              Statistics.Matrix                 Statistics/Matrix.hs:(268,1)-(270,22)                    0.9    1.3
basicUnsafeSlice       Data.Vector.Primitive.Mutable     Data/Vector/Primitive/Mutable.hs:(84,3)-(85,25)          0.6    1.2
```

```     
benchmarking final algebra/1
time                 694.9 ns   (588.1 ns .. 886.2 ns)
                     0.779 R²   (0.705 R² .. 0.994 R²)
mean                 711.0 ns   (626.1 ns .. 914.5 ns)
std dev              394.6 ns   (202.7 ns .. 661.0 ns)
variance introduced by outliers: 100% (severely inflated)
             
benchmarking final algebra/100
time                 55.56 μs   (51.88 μs .. 61.24 μs)
                     0.958 R²   (0.912 R² .. 0.998 R²)
mean                 53.21 μs   (51.88 μs .. 57.81 μs)
std dev              7.194 μs   (2.823 μs .. 14.14 μs)
variance introduced by outliers: 90% (severely inflated)
             
benchmarking final algebra/1000
time                 619.8 μs   (577.1 μs .. 658.1 μs)
                     0.979 R²   (0.971 R² .. 0.993 R²)
mean                 615.3 μs   (598.8 μs .. 634.3 μs)
std dev              63.55 μs   (50.65 μs .. 77.26 μs)
variance introduced by outliers: 76% (severely inflated)
```

### CODE REVIEW

**Comment by eborden** [comment]
This is currently WIP, as `runGraphula_` functions need to be converted and the free representation removed.

**Comment by eborden** [comment]
image

File: graphula-core/src/Graphula.hs [hunk 1]
```diff
@@ -16,6 +16,7 @@
   @
 -}
 {-# LANGUAGE ConstraintKinds #-}
+{-# LANGUAGE ConstrainedClassMethods #-}
```
Comment (cdparks) [inline graphula-core/src/Graphula.hs]: Huh, I didn't know this wasn't Haskell98. There's a pretty funny comment about it [here](https://downloads.haskell.org/~ghc/8.0.2/docs/html/users_guide/glasgow_exts.html#class-method-types):
> The restriction is a pretty stupid one in the first place, so -XConstrainedClassMethods is implied by -XMultiParamTypeClasses.

**Comment by eborden** [comment]
The additional `newtype` wrapping with `ReaderT` comes nowhere near the pain of `Free`

image

File: graphula-persistent/src/Graphula/Persist.hs [hunk 2]
```diff
@@ -1,36 +1,51 @@
...
 import GHC.TypeLits (TypeError, ErrorMessage(..))
 
-persistGraph
-  :: (SqlBackendCanWrite backend, MonadIO m, MonadIO n)
-  => (forall b. ReaderT backend n b -> m b)
-  -> Frontend (PersistRecord backend) Entity (m a) -> m a
-persistGraph runDB = \case
-  Insert n next -> do
-    x <- runDB $ do
+newtype RunDB backend n m = RunDB (forall b. ReaderT backend n b -> m b)
```
Comment (cdparks) [inline graphula-persistent/src/Graphula/Persist.hs]: Curious why we have both `n` and `m` here and below?

Comment (cdparks) [reply graphula-persistent/src/Graphula/Persist.hs]: Oop, never mind, `withGraph` above is a good example. `runDB` might peel off a bunch of nonsense to get to `m b`

File: graphula-core/src/Graphula.hs [hunk 3]
```diff
@@ -86,180 +94,185 @@ import System.Directory (getTemporaryDirectory)
...
+  )
+
+class MonadGraphulaFrontend m where
+  type NodeConstraint m :: * -> Constraint
+  -- ^ A constraint applied to nodes. This is utilized during
+  --   insertion and can be leveraged by frontends with typeclass interfaces
+  --   to insertion.
+  type Node m :: * -> *
+  -- ^ A wrapper type used to return relevant information about a given
+  --   node. `graphula-persistent` returns all nodes in the 'Database.Persist.Entity' type.
```
Comment (cdparks) [inline graphula-core/src/Graphula.hs]: I like that you're giving concrete examples for these

File: graphula-core/src/Graphula.hs [hunk 4]
```diff
@@ -86,180 +94,185 @@ import System.Directory (getTemporaryDirectory)
...
+  --   JSON 'Value's.
+  type Generate m :: * -> Constraint
+  -- ^ A constraint for pluggable node generation. 'runGraphula'
+  --   utilizes 'Arbitrary', 'runGraphulaReplay' utilizes 'FromJSON'.
+  generateNode :: Generate m a => m a
+  logNode :: Logging m a => a -> m ()
+
+newtype GraphulaT m a = GraphulaT
+  { runGraphulaT :: m a
+  -- ^ Run a graphu utilizing 'Arbitrary' for node generation.
```
Comment (cdparks) [inline graphula-core/src/Graphula.hs]: Probably meant `graph`

**Comment by cdparks** [comment]
@eborden generally I think this looks good. It's a little unclear how each flavor of `Graphula` is supposed to be used  together. This is what I was able to gather, but it might be nice to have a central diagram or explanation about how these fit together:
* `GraphulaIdentity` uses `Identity` as the node, but it's carrying around a computation in `IO` (?)
* `GraphulaT` seems to be the main transformer type that must be applied to some concrete type with instances for `MonadGraphulaFrontend/Backend` (e.g. `GraphulaIdentity` or `GraphulaPersistT SqlBackend _ IO`)
* `GraphulaIdempotentT`, `GraphulaLoggedT`, and `GraphulaReplayT` add further functionality to `GraphulaT`

**Comment by eborden** [comment]
* `GraphulaIdentity` uses Identity as the node, but it's carrying around a computation in IO (?)
* `GraphulaT` seems to be the main transformer type that must be applied to some concrete type with instances for MonadGraphulaFrontend/Backend (e.g. `GraphulaIdentity` or `GraphulaPersistT` SqlBackend _ IO`)
* `GraphulaIdempotentT`, `GraphulaLoggedT`, and `GraphulaReplayT` add further functionality to `GraphulaT`

There are seemingly three categories of transformers.
#### Backend
There are concrete backends that implement all of that type classes functionality.
* `GraphulaT` - `Arbitrary` generation with no logging.
* `GraphulaLoggedT` - `Arbitrary` generation with `JSON` logging.
* `GraphulaReplayT` - generation from a log file via `JSON`, no logging.
#### Frontend
* `GraphulaIdentity` - A simple identity wrapper around a generated node.
* `GraphulaPersistT` - Insertion and deletion from a persistent database.
#### Frontend Extensions
* `GraphulaIdempotentT` - Extends a given fronted to run removal finalizers on error or completion.

So you could do:
```
runGraphulaIdentity $ runGraphulaT graph
runGraphulaIdentity . runGraphulaIdemptotentT $ runGraphulaT graph
runGraphulaIdentity $ runGraphulaLoggedT graph
runGraphulaIdentity . runGraphulaIdemptotentT $ runGraphulaLoggedT graph
runGraphulaIdentity $ runGraphulaReplayT logFile graph
runGraphulaIdentity . runGraphulaIdemptotentT $ runGraphulaReplayT logFile graph
runGraphulaPersist runDB $ runGraphulaT graph
runGraphulaPersist runDB . runGraphulaIdemptotentT $ runGraphulaT graph
runGraphulaPersist runDB $ runGraphulaLoggedT graph
runGraphulaPersist runDB . runGraphulaIdemptotentT $ runGraphulaLoggedT graph
runGraphulaPersist runDB $ runGraphulaReplayT logFile graph
runGraphulaPersist runDB . runGraphulaIdemptotentT $ runGraphulaReplayT logFile graph
```

Or you could just implement the frontend/backend type classes for your specific transformer stack and roll your own graphula runner.

**Comment by eborden** [comment]
`GraphulaIdempotentT` could almost be thought of as middleware.

**Pull request merged**